In [7]:
from pathlib import Path
import pandas as pd
import numpy as np

DATA_DIR = Path("../data/raw")

customers = pd.read_csv(DATA_DIR / "olist_customers_dataset.csv")
orders = pd.read_csv(DATA_DIR / "olist_orders_dataset.csv")
order_items = pd.read_csv(DATA_DIR / "olist_order_items_dataset.csv")
payments = pd.read_csv(DATA_DIR / "olist_order_payments_dataset.csv")
reviews = pd.read_csv(DATA_DIR / "olist_order_reviews_dataset.csv")
products = pd.read_csv(DATA_DIR / "olist_products_dataset.csv")
sellers = pd.read_csv(DATA_DIR / "olist_sellers_dataset.csv")
geolocation = pd.read_csv(DATA_DIR / "olist_geolocation_dataset.csv")
category_translation = pd.read_csv(DATA_DIR / "product_category_name_translation.csv")

In [8]:
tables = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "payments": payments,
    "reviews": reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
    "category_translation": category_translation
}

table_summary = pd.DataFrame([
    {
        "table": name,
        "rows": df.shape[0],
        "columns": df.shape[1]
    }
    for name, df in tables.items()
])

table_summary

,table,rows,columns
0,customers,99441,5
1,orders,99441,8
2,order_items,112650,7
3,payments,103886,5
4,reviews,99224,7
5,products,32951,9
6,sellers,3095,4
7,geolocation,1000163,5
8,category_translation,71,2


In [9]:
for name, df in tables.items():
    print(f"\n{'=' * 60}")
    print(f"TABLE: {name}")
    print(f"{'=' * 60}")
    print(df.columns.tolist())


TABLE: customers
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

TABLE: orders
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

TABLE: order_items
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

TABLE: payments
['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

TABLE: reviews
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

TABLE: products
['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

TABLE: sellers
['seller_id', 'seller_zip_code_prefix', 'seller_cit

In [10]:
for name, df in tables.items():
    print(f"\n{'=' * 60}")
    print(f"TABLE: {name}")
    print(f"{'=' * 60}")
    display(df.head(3))


TABLE: customers


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP



TABLE: orders


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00



TABLE: order_items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87



TABLE: payments


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71



TABLE: reviews


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24



TABLE: products


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0



TABLE: sellers


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ



TABLE: geolocation


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP



TABLE: category_translation


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto


In [11]:
for name, df in tables.items():
    print(f"\n{'=' * 60}")
    print(f"TABLE: {name}")
    print(f"{'=' * 60}")
    print(df.dtypes)


TABLE: customers
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

TABLE: orders
order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

TABLE: order_items
order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

TABLE: payments
order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

TABLE: reviews
review_id      

In [12]:
missing_summary = pd.DataFrame([
    {
        "table": name,
        "column": col,
        "missing_count": df[col].isna().sum(),
        "missing_pct": round(df[col].isna().mean() * 100, 2)
    }
    for name, df in tables.items()
    for col in df.columns
    if df[col].isna().sum() > 0
])

missing_summary.sort_values(
    ["missing_pct", "missing_count"],
    ascending=False
)

,table,column,missing_count,missing_pct
3,reviews,review_comment_title,87656,88.34
4,reviews,review_comment_message,58247,58.70
2,orders,order_delivered_customer_date,2965,2.98
5,products,product_category_name,610,1.85
6,products,product_name_lenght,610,1.85
7,products,product_description_lenght,610,1.85
8,products,product_photos_qty,610,1.85
1,orders,order_delivered_carrier_date,1783,1.79
0,orders,order_approved_at,160,0.16
9,products,product_weight_g,2,0.01


## Missing Value Observations

- `review_comment_title` has a very high percentage of missing values (88.34%). This may indicate that many customers did not provide a review title.
- `review_comment_message` has a high percentage of missing values (58.70%). This may indicate that many customers did not provide written comments.
- `order_delivered_customer_date` has 2.98% missing values. The reason for these missing values needs to be investigated using `order_status`.
- `order_delivered_carrier_date` has 1.79% missing values. The reason for these missing values should also be investigated using order status and other order timestamps.
- Several product attributes have 1.85% missing values. These include category, name length, description length, and photo count. The reason for these missing values needs further investigation.
- Product dimension fields have very few missing values (0.01%), so their impact is likely small, but they should still be documented.

In [13]:
duplicate_summary = pd.DataFrame([
    {
        "table": name,
        "duplicate_rows": df.duplicated().sum(),
        "duplicate_pct": round(df.duplicated().mean() * 100, 2)
    }
    for name, df in tables.items()
])

duplicate_summary.sort_values(
    "duplicate_rows",
    ascending=False
)

,table,duplicate_rows,duplicate_pct
7,geolocation,261831,26.18
0,customers,0,0.00
1,orders,0,0.00
2,order_items,0,0.00
3,payments,0,0.00
4,reviews,0,0.00
5,products,0,0.00
6,sellers,0,0.00
8,category_translation,0,0.00


In [15]:
geolocation.groupby(
    [
        "geolocation_zip_code_prefix",
        "geolocation_lat",
        "geolocation_lng",
        "geolocation_city",
        "geolocation_state"
    ]
).size().sort_values(ascending=False).head(10)
print("Unique ZIP prefixes:", geolocation["geolocation_zip_code_prefix"].nunique())

print("\nRecords per ZIP prefix:")
print(
    geolocation.groupby("geolocation_zip_code_prefix")
    .size()
    .describe()
)

Unique ZIP prefixes: 19015

Records per ZIP prefix:
count    19015.000000
mean        52.598633
std         72.057907
min          1.000000
25%         10.000000
50%         29.000000
75%         66.500000
max       1146.000000
dtype: float64


In [16]:
geolocation.duplicated().sum()

np.int64(261831)

## Duplicate Value Observations

- The `geolocation` table contains **261,831 completely duplicated rows**.
- `geolocation_zip_code_prefix` is not unique, with **19,015 unique ZIP prefixes** across the table.
- A single ZIP prefix can therefore have multiple geolocation records.
- The duplicate records will **not be removed at this stage**. Their treatment will be decided during the data cleaning phase after understanding how the table will be used.

In [17]:
key_checks = {
    "customers.customer_id": customers["customer_id"].nunique(),
    "customers.customer_unique_id": customers["customer_unique_id"].nunique(),
    "orders.order_id": orders["order_id"].nunique(),
    "order_items.order_id": order_items["order_id"].nunique(),
    "order_items.product_id": order_items["product_id"].nunique(),
    "sellers.seller_id": sellers["seller_id"].nunique(),
    "products.product_id": products["product_id"].nunique(),
}

key_checks

{'customers.customer_id': 99441,
 'customers.customer_unique_id': 96096,
 'orders.order_id': 99441,
 'order_items.order_id': 98666,
 'order_items.product_id': 32951,
 'sellers.seller_id': 3095,
 'products.product_id': 32951}

In [18]:
print("Customers:")
print("customer_id duplicates:", customers["customer_id"].duplicated().sum())
print("customer_unique_id duplicates:", customers["customer_unique_id"].duplicated().sum())

print("\nOrders:")
print("order_id duplicates:", orders["order_id"].duplicated().sum())

print("\nProducts:")
print("product_id duplicates:", products["product_id"].duplicated().sum())

print("\nSellers:")
print("seller_id duplicates:", sellers["seller_id"].duplicated().sum())

Customers:
customer_id duplicates: 0
customer_unique_id duplicates: 3345

Orders:
order_id duplicates: 0

Products:
product_id duplicates: 0

Sellers:
seller_id duplicates: 0
